marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Api.hs

```diff
@@ -0,0 +1,384 @@
+  , metadata      :: !(Maybe Metadata)       -- ^ The metadata of this transaction (only includes Marlowe-relevant metadata).
+  , inputs        :: !(Set TransactionInput) -- ^ The inputs consumed by the transaction
+  , outputs       :: ![TransactionOutput]    -- ^ The outputs produced by the transaction.
+  , mintedTokens  :: !Tokens                 -- ^ Tokens minted by the transaction.
+  }
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving anyclass (Binary)
+
+-- | A validity range for a transaction
+data ValidityRange
+  = Unbounded                 -- ^ The transaction is always valid.
+  | MinBound SlotNo           -- ^ The transaction is only valid after a specific slot.
+  | MaxBound SlotNo           -- ^ The transaction is only valid before a specific slot.
+  | MinMaxBound SlotNo SlotNo -- ^ The transaction is only valid between two slots.
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving anyclass (Binary)
+
+-- TODO invlude validator versions
+data Metadata
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving anyclass (Binary)
+
+-- | An input of a transaction.
+data TransactionInput = TransactionInput
+  { txId     :: !TxId             -- ^ The txId of the TransactionOutput this input consumes.
+  , txIx     :: !TxIx             -- ^ The txIx of the TransactionOutput this input consumes.
+  , redeemer :: !(Maybe Redeemer) -- ^ The script redeemer dataum for this input (if one was provided).
+  }
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving anyclass (Binary)
+
+-- | An output of a transaction.
+data TransactionOutput = TransactionOutput
+  { address   :: !Address           -- ^ The address that receives the assets of this output.
+  , assets    :: !Assets            -- ^ The assets this ouptut produces.
+  , datumHash :: !(Maybe DatumHash) -- ^ The hash of the script datum associated with this output.
+  , datum     :: !(Maybe Datum)     -- ^ The script datum associated with this output.
+  }
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving anyclass (Binary)
+
+-- | A script datum that is used to spend the output of a script tx.
+newtype Redeemer = Redeemer { unRedeemer :: Datum }
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving newtype (Binary)
+
+-- | A datum as a sum-of-products.
+data Datum
+  = Constr Integer [Datum]
+  | Map [(Datum, Datum)]
+  | List [Datum]
+  | I Integer
+  | B ByteString
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving anyclass (Binary)
+
+-- | Convert from Plutus.V1.Ledger.Api.Data to Datum
+fromPlutusData :: Plutus.Data -> Datum
+fromPlutusData (Plutus.Constr i dats) = Constr i $ fromPlutusData <$> dats
+fromPlutusData (Plutus.Map m)         = Map $ bimap fromPlutusData fromPlutusData <$> m
+fromPlutusData (Plutus.List dats)     = List $ fromPlutusData <$> dats
+fromPlutusData (Plutus.I i)           = I i
+fromPlutusData (Plutus.B b)           = B b
+
```

Comment:
Will we need `toPlutusData`, too? If so, we should include it here instead of having to create it in another package that depends upon this module.

Reply:
Yes, this will very likely be needed.